## Assignment Set 1: Classification and Regression
---
> COMPILED BY NAKACWA OLIVIA 20/U/0080




## Question 1: Classification (Topics: Classification, Model Evaluation)

In [ ]:
# Import necessary libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [ ]:
# Load data
data = pd.read_csv('diabetes.csv')

In [ ]:
# Displaying the first five rows
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
# Checking empty columns
data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [ ]:
# Converting the data to dataframe
df = pd.DataFrame(data)
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [ ]:
# Checking the information of the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [ ]:
# Create feature matrix and target vector
X = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

In [ ]:
from sklearn.model_selection import train_test_split
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [ ]:
# View train data
print(X_train)

[[  9.     89.     62.    ...  22.5     0.142  33.   ]
 [  1.    118.     58.    ...  33.3     0.261  23.   ]
 [  0.     91.     80.    ...  32.4     0.601  27.   ]
 ...
 [  4.     94.     65.    ...  24.7     0.148  21.   ]
 [ 11.     85.     74.    ...  30.1     0.3    35.   ]
 [  5.    136.     82.    ...   0.      0.64   69.   ]]


In [ ]:
# View train labels
print(y_train)

[0 0 0 1 1 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 1 1 0 0 1 0 1 1 0 0 1 1 0
 0 0 1 0 0 1 1 0 1 0 1 0 1 0 1 0 1 0 0 0 0 0 1 0 1 0 1 0 0 0 0 0 0 0 0 1 0
 1 1 1 1 0 0 0 0 1 1 0 1 0 0 1 1 0 0 0 0 0 0 0 1 1 0 0 0 0 1 1 0 0 0 0 1 0
 0 1 1 0 1 1 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 1 1 1 1 0 0 0 0 0 1 0
 0 0 0 1 1 0 0 0 1 1 1 1 0 1 0 1 1 0 0 0 1 0 0 0 1 0 1 1 0 1 1 0 0 0 0 1 1
 0 0 1 1 0 0 0 0 0 1 0 1 0 0 0 1 0 1 0 1 0 1 1 1 1 0 1 1 0 0 1 1 0 0 1 1 0
 0 0 1 1 0 0 1 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0
 0 0 0 0 1 1 0 1 0 1 1 1 0 1 0 0 1 1 0 0 0 1 0 0 1 0 1 0 0 0 0 1 0 0 0 1 1
 0 1 0 1 1 0 1 0 1 0 0 0 1 0 1 0 1 0 0 1 1 0 0 0 0 1 0 0 1 0 1 1 0 0 1 1 0
 0 0 0 1 0 0 0 0 1 0 1 0 0 0 0 1 0 0 0 1 0 0 0 1 1 0 0 0 0 0 1 0 0 1 0 1 1
 0 1 0 0 0 0 1 1 0 0 0 1 0 0 0 0 1 0 1 0 0 0 1 0 1 1 1 1 0 1 0 0 0 0 0 1 0
 1 0 1 0 0 0 1 1 0 0 1 1 0 1 0 0 1 0 0 1 0 0 0 0 0 1 1 1 1 0 0 1 1 0 0 0 1
 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1
 1 0 1 1 0 0 1 0 0 0 1 0 

In [ ]:
from sklearn.preprocessing import StandardScaler
# Perform feature scaling on the train and test data
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [ ]:
# Displaying the train data
X_train

array([[ 1.50755225, -1.01521454, -0.40451932, ..., -1.22070104,
        -0.98325882, -0.04863985],
       [-0.82986389, -0.09964691, -0.61509602, ...,  0.13719053,
        -0.62493647, -0.88246592],
       [-1.12204091, -0.95207195,  0.54307587, ...,  0.0240329 ,
         0.39884168, -0.5489355 ],
       ...,
       [ 0.04666716, -0.85735805, -0.24658679, ..., -0.9440935 ,
        -0.96519215, -1.04923114],
       [ 2.09190629, -1.14149973,  0.2272108 , ..., -0.26514771,
        -0.5075031 ,  0.11812536],
       [ 0.33884418,  0.46863645,  0.64836422, ..., -4.04964181,
         0.51627505,  2.953134  ]])

In [ ]:
# Displaying the test data
X_test

array([[-0.82986389,  2.4576282 ,  0.33249916, ...,  1.34420526,
         2.78665365, -0.96584853],
       [-0.53768687, -0.44693118,  0.2272108 , ...,  0.17490974,
        -0.19434743, -0.88246592],
       [ 0.04666716, -1.42564141, -0.40451932, ...,  0.22520202,
        -0.23349189, -0.71570071],
       ...,
       [-0.82986389, -0.38378859, -0.50980767, ..., -0.85608201,
         1.4406865 , -1.04923114],
       [-0.24550986,  0.21606607,  0.43778751, ..., -1.39672402,
        -0.60385869,  1.7857775 ],
       [ 0.33884418, -1.04678584,  0.43778751, ..., -0.57947447,
        -0.63396981,  0.28489057]])

### Using Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
# Fit Logistic Regression model
classifier = LogisticRegression(random_state=42)
classifier.fit(X_train,y_train)

LogisticRegression(random_state=42)

In [ ]:
# Predict test set results
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 1]
 [1 0]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
# Making the Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
print('Confusion Matrix :\n',cm)
# Displaying Accuracy Score
acc = accuracy_score(y_test,y_pred)
print('Accuracy :',acc)

Confusion Matrix :
 [[118  12]
 [ 26  36]]
Accuracy : 0.8020833333333334


In [ ]:
# Displaying the classification report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.91      0.86       130
           1       0.75      0.58      0.65        62

    accuracy                           0.80       192
   macro avg       0.78      0.74      0.76       192
weighted avg       0.80      0.80      0.79       192



### Using Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier
# Fit Decision Tree model
tree_classifier = DecisionTreeClassifier(criterion='entropy',random_state=42)
tree_classifier.fit(X_train,y_train)

DecisionTreeClassifier(criterion='entropy', random_state=42)

In [ ]:
# Predict test set results
y_pred = tree_classifier.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
# Displaying Accuracy Score
acc = accuracy_score(y_test,y_pred)
print('Accuracy :',acc)

Accuracy : 0.78125


### Comparison of the two models

### The Logistic Regression model has given a better accuracy value of 0.80208 whereas the decision tree model whose accuracy value is 0.78125

### Strengths of the models

#### Logistic Regression

*   It is a straightforward algorithm that provides interpretable results.
*   It handles small sample sizes and still provide reliable estimates. It is a suitable choice when the available data is limited.
*   It accommodates both categorical and continuous predictor variables.
*   It produces probability estimates for each class, allowing for the assessment of confidence in predictions.
*   It is relatively robust to noise and outliers in the data.

#### Decision Tree
*   It captures nonlinear relationships between predictors and the outcome.
*   It naturally captures interactions between predictors.
*   It provides a clear and intuitive representation of decision-making.
*   It handles both categorical and continuous variables without requiring explicit encoding.
*   It is less sensitive to outliers compared to some other algorithms.
*   It does not assume a particular distribution of the data, making them suitable for a wide range of data types and distributions.
*   It handles missing values by determining the optimal splits based on available data for each predictor.



### Weaknesses of the models

#### Logistic Regression
*   It assumes a linear relationship between predictors and the log-odds of the outcome.
*   It is affected by multicollinearity, which occurs when predictor variables are highly correlated.
*   It assumes that observations are independent of each other.
*   It works with small sample sizes, it still requires an adequate number of observations to produce reliable estimates and avoid overfitting.



#### Decision Tree
*   It is prone to overfitting, especially when they are allowed to grow without constraints.
*   It is sensitive to small changes in the data.
*   Each split in a decision tree is locally optimal at the time it is made. However, these local decisions may not lead to the best overall tree structure, as the algorithm does not consider the global optimization of the tree.
*   It tends to favor variables with more categories or levels in their splitting decisions.
*   It struggles to capture certain types of relationships, particularly those that require complex interactions or long sequences of decisions.



### Performing hyperparameter tuning on decision tree

In [ ]:
# Import GridSearchCV
from sklearn.model_selection import GridSearchCV
# Defining hyperparameters
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
}

In [ ]:
# Performing a grid search to find the best combination of hyperparameters
grid_search = GridSearchCV(tree_classifier, param_grid, cv=5)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=DecisionTreeClassifier(criterion='entropy',
                                              random_state=42),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [None, 5, 10, 15],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10]})

In [ ]:
# Retrieving the best hyperparameters and the best model
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

In [ ]:
# Evaluating the best model on the testing set
y_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.765625


#### Performing hyperparameter on decision tree model doesnot improve the performance of the model

In [ ]:
# Making the Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
print('Confusion Matrix :\n',cm)
print("Accuracy of the best model: ", acc)

Confusion Matrix :
 [[105  25]
 [ 17  45]]
Accuracy of the best model:  0.78125


# Question 2: Regression (Topics: Regression, Feature Engineering)

In [ ]:
# Import necessary libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [ ]:
# Load data
dataset = pd.read_csv('diabetes.csv')

In [ ]:
# Displaying the dataset
dataset

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [ ]:
# Checking the datatypes of the data
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
dataset.describe()

In [ ]:
# Create feature matrix and target vector
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [ ]:
# Displaying x values
X

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [ ]:
# Displaying y values
y

array([1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,

In [ ]:
# Split data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

## Simple linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression
# Instantiate the model
model_regressor = LinearRegression()
#Fit simple Linear Regression model
model_regressor.fit(X_train, y_train)

LinearRegression()

In [ ]:
# Predict test set results
y_pred = model_regressor.predict(X_test)

In [ ]:
# Calculating and displaying the mean absolute error
from sklearn.metrics import mean_absolute_error
print("Mean Absolute Error: ",mean_absolute_error(y_test,y_pred))

Mean Absolute Error:  0.34812825999928243


In [ ]:
# Calculating and displaying the mean squared error
from sklearn.metrics import mean_squared_error
print("Mean Squared Error: ",mean_squared_error(y_test,y_pred))

Mean Squared Error:  0.171045272808501


In [ ]:
# Calculating and displaying the root mean squared error
print("Root Mean Squared Error: ",np.sqrt(mean_squared_error(y_test,y_pred)))

Root Mean Squared Error:  0.41357619951890484


In [ ]:
# Calculating and displaying RMSLE
print("Root Mean Squared Log Error: ",np.log(np.sqrt(mean_squared_error(y_test,y_pred))))

Root Mean Squared Log Error:  -0.8829135021271755


In [ ]:
# Calculating and displaying the R2 score
from sklearn.metrics import r2_score
print("R2",r2_score(y_test,y_pred))

R2 0.2550028117674177


In [ ]:
# Interpreting the results
coefficients = model_regressor.coef_
intercept = model_regressor.intercept_
print("Coefficients:", coefficients)
print("Intercept:", intercept)

Coefficients: [ 0.01046818  0.00563353 -0.00228103  0.00053209 -0.00027815  0.01503864
  0.11126264  0.00646027]
Intercept: -0.9487546338208501


## Gradient Boosting Regression

In [ ]:
# Importing the necessary libraries
from sklearn.ensemble import GradientBoostingRegressor
# Instantiating the model
model_gradient = GradientBoostingRegressor(n_estimators=100, random_state=42)
# Fitting the model
model_gradient.fit(X_train, y_train)

GradientBoostingRegressor(random_state=42)

In [ ]:
# Making predictions on the test set
y_pred = model_gradient.predict(X_test)

In [ ]:
# # Calculating and displaying the mean absolute error
from sklearn.metrics import mean_absolute_error
print("Mean Absolute Error: ",mean_absolute_error(y_test,y_pred))

Mean Absolute Error:  0.3226644258448055


In [ ]:
# Calculating and displaying the mean squared error
from sklearn.metrics import mean_squared_error
print("Mean Squared Error: ",mean_squared_error(y_test,y_pred))

Mean Squared Error:  0.17341954529772322


In [ ]:
# Calculating and displaying the root mean squared error
print("Root Mean Squared Error: ",np.sqrt(mean_squared_error(y_test,y_pred)))

Root Mean Squared Error:  0.416436724242379


In [ ]:
# Calculating and displaying the MSLE
print("Root Mean Squared Log Error: ",np.log(np.sqrt(mean_squared_error(y_test,y_pred))))

Root Mean Squared Log Error:  -0.8760207515039703


In [ ]:
# Calculating and displaying the R2 score
from sklearn.metrics import r2_score
print("R2",r2_score(y_test,y_pred))

R2 0.24466153603658314


## Comparison the two models
#### The grandient boosting algorithm has a low R2 score as compared to Linear Regression

## Strengths of the models

### SImple Linear Regression

*   It is a simple and transparent model that provides easy interpretability
*   It  assumes a linear relationship between the independent variables and the dependent variable.
*   It has low computational complexity, making it computationally efficient for large datasets.
*   It can be extended to handle more complex relationships by incorporating polynomial terms, interaction terms, or using transformations of variables.




### Gradient Boosting Regression


*   It captures complex non-linear relationships between variables and effectively model intricate patterns in the data, leading to accurate predictions.
*   It can handle a variety of data types, including both numerical and categorical features.
*   It provides a measure of feature importance, allowing you to assess the contribution of each feature to the prediction.
*   It handles imbalanced datasets effectively.
*   It provides regularization techniques such as learning rate and tree depth, which can help prevent overfitting.

## Weaknesses of the models

### Simple Linear Regression

*   It is limited to modeling linear relationships.
*   It assumes that the observations are independent and have constant variance (homoscedasticity).
*   It requires categorical variables to be transformed into numerical representations, which can result in a large number of dummy variables.
*   It assumes that the errors (residuals) follow a normal distribution.



### Gradient Boosting Regression

*   It is computationally expensive and time-consuming, especially when dealing with large datasets or complex models.
*   It has several hyperparameters that need to be carefully tuned to achieve optimal performance.
*   It is prone to overfitting, especially when the model becomes too complex or when hyperparameters are not properly tuned.
*   It is sensitive to noisy data, outliers, or mislabeled data points.
*   It requires categorical variables to be one-hot encoded, resulting in an expanded feature space.

## Future Engineering techniques

In [ ]:
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
# Performing feature transformation
polynomial_features = PolynomialFeatures(degree=2)
X_train_poly = polynomial_features.fit_transform(X_train)
X_test_poly = polynomial_features.transform(X_test)

In [ ]:
# Standardising the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_poly)
X_test_scaled = scaler.transform(X_test_poly)

In [ ]:
# Training the model
regressor = LinearRegression()
regressor.fit(X_train_scaled, y_train)

LinearRegression()

In [ ]:
# Evaluating the model
y_pred = regressor.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
print("Mean Squared Error: ",mse)

Mean Squared Error:  0.1788041440297847
